### ISSUE TITLE
Getting segmentation fault for custom loss

### ISSUE THREAD

**Issue by frd-glovo** [microsoft/LightGBM#2750]
Hi,

I've been trying the following custom loss for a regression task, and I'm getting a segmentation fault:

```
def hm(y_true, y_pred):
    residual = (y_true - y_pred).astype("float")
    residual_abs = np.fabs(residual)
    grad = np.where(residual_abs<=300, 
                    np.copysign(1, -residual)* 0.5,
                    np.where(residual_abs>=900, 
                             np.copysign(1, -residual)* 1,
                             np.copysign(1, -residual)* 2))
    hess = np.ones(y_true.shape[0]) 
    return grad, hess

```
Any ideas why it might be failing? Im using goss and default hyperparameters

**Comment by guolinke** [comment]
Which code version you use?

**Comment by frd-glovo** [comment]
I was initially using lightgbm==2.2.2. After updating to latest, I still get the seg fault.

By the way, with this loss function, will the leaf outputs be the median (or some other similar value)  of the residuals, regardless of the boosting type?  Or will it be L'^2/L'' ?  I read your last post [here](https://github.com/microsoft/LightGBM/issues/2128) and it's not clear to me if lightgbm will automatically detect constant hessian and start calculating leaf outputs differently.

**Comment by frd-glovo** [comment]
I've also just realized that the KDE bandwith error is not an error during training, but something triggered later as a result of the predicted values. I will correct the OP.

**Comment by guolinke** [comment]
could you provide a reproduce example? if your data cannot be posted here, you can try the randomly generated data.

For the custom obj, LightGBM will treat it as non-constant hessian (of course, it could be improved by check L'' before train). And the output for custom obj will always be ` L'/L'', not the median of residuals.

**Comment by frd-glovo** [comment]
Sure, I will try to prepare a reproducible sample!

But being ` L'/L'', the output for custom loss functions, writing constant gradient losses doesn't seem very useful. We will try to rewrite this loss function to accomplish the same objective using a Loss function with non constant hessian. 

Does lightgbm use the second order approximation for the "already" implemented L1 losses? I've been trying to wrap my head around how L1 losses are handled and I'm confused. I've asked about this on stack overflow (https://stackoverflow.com/questions/60137453/how-does-lightgbm-or-other-boosted-trees-implementations-with-2nd-order-approxi) maybe someone from the maintainers can give a hand?

**Comment by guolinke** [comment]
refer to code of built-in L1 objective  
https://github.com/microsoft/LightGBM/blob/d8a34df92d996ab82b56d29188210e9eef1f80cb/src/objective/regression_objective.hpp#L189-L270

Before this version, we use the second-order approximation, but its performance actually is not good. 
And we switch back to 1) use first-order gradient to find split point; 2) then use the median of residuals for leaf outputs, as shown in the above code.

**Comment by StrikerRUS** [comment]
ping @frd-glovo for a repro

> Sure, I will try to prepare a reproducible sample!
